### PR TITLE
Configuration of Implementation Model

### PR DESCRIPTION
**Pull request by gkahiu**
Mapping of NCS pathways to implementation models in step two of setting up a scenario.

### CODE REVIEW

**Comment by gkahiu** [comment]
@Samweli please see the near-final changes to step two, only thing remaining is defining the file names for the NCS pathway files [here](https://github.com/kartoza/cplus-plugin/pull/137/files#diff-d0aa74f9c7c1a5810153eed56e37c4b60098d5f556f14ada1bb0db6b3630ae65R20), could not update as am unable to download the ZIP files, will follow-up separately. Also, this will also be dependent on the settings work, specifically on the base data directory so that we can use this to build the absolute path for the NCS pathway layers using [this](https://github.com/kartoza/cplus-plugin/pull/137/files#diff-2614dd89bd12be5f3a169c76a20ed63c2f409e15f136d5729bbadb34c12cde94R441) snippet.

**Comment by Samweli** [comment]
> @Samweli please see the near-final changes to step two, only thing remaining is defining the file names for the NCS pathway

Thanks @gkahiu, can you fix the PR conflicts then I can post my review and we can move forward from there, thanks.

File: src/cplus_plugin/gui/implementation_model_widget.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+class ImplementationModelContainerWidget(QtWidgets.QWidget, WidgetUi):
+    """Widget for configuring the implementation model."""
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        self.setupUi(self)
+
+        self._items_loaded = False
+
+        self.btn_add_one.setIcon(FileUtils.get_icon("play.svg"))
```
Comment (Samweli) [inline src/cplus_plugin/gui/implementation_model_widget.py:41]: QGIS theme icons might be a better option here eg. https://github.com/qgis/QGIS/blob/master/images/themes/default/mActionPlay.svg and https://github.com/qgis/QGIS/tree/master/images/themes/default/temporal_navigation

File: admin.py [hunk 2]
```diff
@@ -281,7 +281,9 @@ def compile_resources(
     target_path = output_directory / "resources.py"
     target_path.parent.mkdir(parents=True, exist_ok=True)
     _log(f"compile_resources target_path: {target_path}", context=context)
-    subprocess.run(shlex.split(f"pyrcc5 -o {target_path} {resources_path}"))
+    subprocess.run(
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply admin.py:284]: These was a temporary solution while awaiting approval of the previous PR on admin commands. Will retract this and use what is currently in the `main` branch.

File: src/cplus_plugin/models/base.py [hunk 3]
```diff
@@ -19,12 +26,246 @@ class SpatialExtent:
...
+        :rtype: bool
+        """
+        layer = self.to_map_layer()
+        if layer is None:
+            return False
+
+        return layer.isValid()
+
+
+def model_component_to_dict(model_component: BaseModelComponentType) -> dict:
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/models/base.py]: This is now effected.

File: src/cplus_plugin/conf.py [hunk 4]
```diff
@@ -259,5 +276,188 @@ def delete_all_scenarios(self):
...
+                if implementation_model is not None:
+                    implementation_models.append(implementation_model)
+
+        return sorted(implementation_models, key=lambda imp_model: imp_model.name)
+
 
 settings_manager = SettingsManager()
+
+
+def initialize_default_settings():
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/conf.py]: This has been effected.

File: src/cplus_plugin/gui/qgis_cplus_main.py [hunk 5]
```diff
@@ -53,6 +55,12 @@ def __init__(
         self.setupUi(self)
         self.iface = iface
 
+        # Insert widget for step 2
+        self.implementation_model_widget = ImplementationModelContainerWidget(self)
+        self.tabWidget.insertTab(1, self.implementation_model_widget, self.tr("Step 2"))
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/qgis_cplus_main.py]: This is now effected.

File: src/cplus_plugin/gui/qgis_cplus_main.py [hunk 6]
```diff
@@ -53,6 +55,12 @@ def __init__(
...
+
+        self.tabWidget.currentChanged.connect(self.on_tab_step_changed)
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/qgis_cplus_main.py]: This is now effected.

File: src/cplus_plugin/gui/ncs_pathway_editor_dialog.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        self.rb_upload.toggled.connect(self._on_upload_from_file)
+        self.buttonBox.accepted.connect(self._on_accepted)
+
+        self._edit_mode = False
+        self._ncs_pathway = ncs_pathway
+        if self._ncs_pathway is not None:
+            self._update_controls()
+            self._edit_mode = True
+
+        help_icon = QgsApplication.instance().getThemeIcon("mActionHelpContents.svg")
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/ncs_pathway_editor_dialog.py:39]: This is now effected..

File: src/cplus_plugin/gui/model_component_widget.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+
+    def __init__(self, parent=None, item_model=None):
+        super().__init__(parent)
+        self.setupUi(self)
+
+        self._item_model = item_model
+        if self._item_model is not None:
+            self.item_model = self._item_model
+
+        add_icon = QgsApplication.instance().getThemeIcon("symbologyAdd.svg")
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/model_component_widget.py]: This is now effected.

File: src/cplus_plugin/gui/model_component_widget.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        self.btn_add.setIcon(add_icon)
+        self.btn_add.clicked.connect(self._on_add_item)
+
+        remove_icon = QgsApplication.instance().getThemeIcon("symbologyRemove.svg")
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/model_component_widget.py]: This is now effected.

File: src/cplus_plugin/gui/model_component_widget.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        self.btn_remove.setIcon(remove_icon)
+        self.btn_remove.setEnabled(False)
+        self.btn_remove.clicked.connect(self._on_remove_item)
+
+        edit_icon = QgsApplication.instance().getThemeIcon("mActionToggleEditing.svg")
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/gui/model_component_widget.py]: This is now effected.

File: src/cplus_plugin/gui/implementation_model_widget.py (hunk 1 above)
Comment (gkahiu) [reply src/cplus_plugin/gui/implementation_model_widget.py]: This is now effected.

**Comment by gkahiu** [comment]
@Samweli see the updated PR which resolves the comments you raised.

On the use of sqlite, I debated on this as well. On one hand, its use is definitely relevant if we are storing multiple scenarios with each scenario being linked to the respective implementation models, which are then subsequently linked to the respective NCS pathways. However, since we, at least for the current implementation, are only dealing with one scenario at any one time then the level of effort vs. the comparative advantage would be a bit skewed. I also considered JSON as another data source/store and maybe we might need to implement some aspects of it in the next epic.

Other considerations included:
- Consistency in using one store for the application data. Hence, the settings approach seemed more appropriate;
- Ease of use in making direct changes - while the `dbmanager` could have been used for editing the NCS pathway properties, the current implementation of settings is much more easier since any text editor can be used to make changes.

That said, the implementation is flexible and can always revert to the original idea of adopting sqlite as the requirements evolve.

File: src/cplus_plugin/gui/ncs_pathway_editor_dialog.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Samweli) [reply src/cplus_plugin/gui/ncs_pathway_editor_dialog.py:39]: @gkahiu thanks, the line above should now change to fetch the icon from the plugin icon directory right?

**Review by Samweli — CHANGES_REQUESTED** [review]

**Comment by gkahiu** [comment]
Hi @Samweli, thanks. The icon source is now fixed.

**Pull request merged**